### ISSUE TITLE
the serial port stops working after a few characters

### ISSUE THREAD

**Issue by snickerbockers** [washingtondc-emu/washingtondc#20]
When attaching to the serial port over telnet, it will print a few characters and then not print any more after that.

This bug was apparently introduced in 93723c6d87167089c3e03abd0daf161113eacbfe way back on August 28.  For the past couple of months I've been testing primarily on official Dreamcast releases (which don't use the serial port for logging) instead of homebrew, which is why I haven't noticed this until now.

**Comment by snickerbockers** [comment]
This will fix it, although I'm not sure I like the solution.  Ideally there should be a a way to raise this interrupt from the io thread without resorting to polling or potentially blocking on a mutex, IDK.

As for the root-cause, I don't know exactly why 10/second is causing troubles.  *Maybe* there's some code in KallistiOS that times out or *maybe* there's some serial signal I'm supposed to raise when transmitting is slow...

`
diff --git a/src/dreamcast.c b/src/dreamcast.c
index a5aa0f5..2ce74ae 100644
--- a/src/dreamcast.c
+++ b/src/dreamcast.c
@@ -111,7 +111,7 @@ static void dreamcast_enable_serial_server(void);
 
 static void dreamcast_enable_cmd_tcp(void);
 
-#define DC_PERIODIC_EVENT_PERIOD (SCHED_FREQUENCY / 10)
+#define DC_PERIODIC_EVENT_PERIOD (SCHED_FREQUENCY / 100)
 static void periodic_event_handler(struct SchedEvent *event);
 static struct SchedEvent periodic_event;
`

**Comment by snickerbockers** [comment]
Applying the following patch to KallistiOS will resolve the problem even without changing DC_PERIODIC_EVENT_PERIOD in WashingtonDC.  This proves that the root-cause of this bug is that there's a timeout in KallistiOS' scif code, and setting the DC_PERIODIC_EVENT_PERIOD to (SCHED_FREQUENCY / 10) causes the scif to respond so slowly that KallistiOS does timeout.  Therefore, the above fix (setting the DC_PERIODIC_EVENT_PERIOD to (SCHED_FREQUENCY / 100)) is appropriate because it does resolve the root-cause of this bug.

```
diff --git a/kernel/arch/dreamcast/hardware/scif.c b/kernel/arch/dreamcast/hardware/scif.c
index 9df880e..6f2eda5 100644
--- a/kernel/arch/dreamcast/hardware/scif.c
+++ b/kernel/arch/dreamcast/hardware/scif.c
@@ -305,7 +305,7 @@ int scif_read() {
 
 /* Write one char to the serial port (call serial_flush()!) */
 int scif_write(int c) {
-    int timeout = 800000;
+    /* int timeout = 800000; */
 
     if(!serial_enabled) {
         errno = EIO;
@@ -314,14 +314,15 @@ int scif_write(int c) {
 
     /* Wait until the transmit buffer has space. Too long of a failure
        is indicative of no serial cable. */
-    while(!(SCFSR2 & 0x20) && timeout > 0)
-        timeout--;
+    while(!(SCFSR2 & 0x20) /* && timeout > 0 */)
+        ;
+        /* timeout--; */
 
-    if(timeout <= 0) {
-        serial_enabled = 0;
-        errno = EIO;
-        return -1;
-    }
+    /* if(timeout <= 0) { */
+    /*     serial_enabled = 0; */
+    /*     errno = EIO; */
+    /*     return -1; */
+    /* } */
 
     /* Send the char */
     SCFTDR2 = c;
@@ -334,7 +335,7 @@ int scif_write(int c) {
 
 /* Flush all FIFO'd bytes out of the serial port buffer */
 int scif_flush() {
-    int timeout = 800000;
+    /* int timeout = 800000; */
 
     if(!serial_enabled) {
         errno = EIO;
@@ -343,14 +344,15 @@ int scif_flush() {
 
     SCFSR2 &= 0xbf;
 
-    while(!(SCFSR2 & 0x40) && timeout > 0)
-        timeout--;
+    while(!(SCFSR2 & 0x40) /* && timeout > 0 */)
+        /* timeout--; */
+        ;
 
-    if(timeout <= 0) {
-        serial_enabled = 0;
-        errno = EIO;
-        return -1;
-    }
+    /* if(timeout <= 0) { */
+    /*     serial_enabled = 0; */
+    /*     errno = EIO; */
+    /*     return -1; */
+    /* } */
 
     SCFSR2 &= 0xbf;
 
```

**Comment by snickerbockers** [comment]
This issue has been resolved by commit 4465b63d8e24c2f2f47749d209b121a137247291.
